java-frontend/src/main/java/org/sonar/java/se/checks/CustomUnclosedResourcesCheck.java

```diff
@@ -0,0 +1,453 @@
+
+  boolean isOpeningResource(MethodInvocationTree syntaxNode) {
+    for (MethodMatcher matcher : openingMethods()) {
+      if (matcher.matches(syntaxNode)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  List<ConstructorMatcher> constructorClasses() {
+    if (classConstructor == null) {
+      if (constructor.length() > 0) {
+        classConstructor = ImmutableList.of(new ConstructorMatcher(constructor));
+      } else {
+        classConstructor = ImmutableList.of();
+      }
+    }
+    return classConstructor;
+  }
+
+  List<MethodMatcher> factoryMethods() {
+    if (factoryList == null) {
+      factoryList = new ArrayList<>();
+      if (factoryMethod.length() > 0) {
+        factoryList = ImmutableList.of(new MethodMatcher(factoryMethod));
+      } else {
+        factoryList = ImmutableList.of();
+      }
+    }
+    return factoryList;
+  }
+
+  List<MethodMatcher> openingMethods() {
+    if (openingList == null) {
+      if (openingMethod.length() > 0) {
+        openingList = ImmutableList.of(new MethodMatcher(openingMethod));
+      } else {
+        openingList = ImmutableList.of();
+      }
+    }
+    return openingList;
+  }
+
+  List<MethodMatcher> closingMethods() {
+    if (closingList == null) {
+      if (closingMethod.length() > 0) {
+        closingList = ImmutableList.of(new MethodMatcher(closingMethod));
+      } else {
+        closingList = ImmutableList.of();
+      }
+    }
+    return closingList;
+  }
+
+  private static abstract class AbstractStatementVisitor extends CheckerTreeNodeVisitor {
+
+    protected AbstractStatementVisitor(ProgramState programState) {
+      super(programState);
+    }
+
+    protected void closeResource(@Nullable final SymbolicValue target) {
+      if (target != null) {
+        ObjectConstraint oConstraint = ObjectConstraint.constraintWithStatus(programState, target, Status.OPENED);
+        if (oConstraint != null) {
+          programState = programState.addConstraint(target.wrappedValue(), oConstraint.withStatus(Status.CLOSED));
+        }
+      }
+    }
+  }
+
+  private class PreStatementVisitor extends AbstractStatementVisitor {
+
+    protected PreStatementVisitor(CheckerContext context) {
+      super(context.getState());
+    }
+
+    @Override
+    public void visitNewClass(NewClassTree syntaxNode) {
+      closeArguments(syntaxNode.arguments(), 0);
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree syntaxNode) {
+      if (isOpeningResource(syntaxNode)) {
+        final SymbolicValue instanceValue = programState.peekValue();
+        programState = programState.addConstraint(instanceValue, new ObjectConstraint(false, false, syntaxNode, Status.OPENED));
+      } else {
+        closeArguments(syntaxNode.arguments(), 1);
+      }
+    }
+
+    @Override
+    public void visitReturnStatement(ReturnStatementTree syntaxNode) {
+      SymbolicValue currentVal = programState.peekValue();
+      if (currentVal != null) {
```

Comment:
How can currentVal be null ? if it is we have some other problems, I don't think this check is necessary.